capsules/extra/src/date_time.rs

```diff
@@ -0,0 +1,354 @@
+
+    fn date_from_u32_tuple(&self, date: u32, time: u32) -> Result<date_time::Date, ErrorCode> {
+        let month_num = date % (1 << 9) / (1 << 5);
+        let month_name = self.u32_as_month(month_num)?;
+
+        let dotw_num = time % (1 << 20) / (1 << 17);
+        let dotw_name = self.u32_as_dotw(dotw_num)?;
+
+        let date_result = date_time::Date {
+            year: (date % (1 << 21) / (1 << 9)) as u16,
+            month: month_name,
+            day: (date % (1 << 5)) as u8,
+
+            day_of_week: dotw_name,
+            hour: (time % (1 << 17) / (1 << 12)) as u8,
+            minute: (time % (1 << 12) / (1 << 6)) as u8,
+            seconds: (time % (1 << 6)) as u8,
+        };
+
+        if !(date_result.day <= 31) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.hour <= 24) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.minute <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.seconds <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+
+        Ok(date_result)
+    }
+
+    /// Transforms Date structure (year, month, dotm, dotw, hour, minute, seconds) into two u32 numbers
+    ///         first number (year, month, day_of_the_month):
+    ///                 -last 5 bits store the day_of_the_month
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
+    ///
+    /// The two u32 numbers are returned as a tuple
+
+    fn date_as_u32_tuple(&self, set_date: date_time::Date) -> Result<(u32, u32), ErrorCode> {
+        let month = self.month_as_u32(set_date.month);
+        let dotw = self.dotw_as_u32(set_date.day_of_week);
+
+        let date = set_date.year as u32 * (1 << 9) as u32
+            + month as u32 * (1 << 5) as u32
+            + set_date.day as u32;
+        let time = dotw as u32 * (1 << 17) as u32
+            + set_date.hour as u32 * (1 << 12) as u32
+            + set_date.minute as u32 * (1 << 6) as u32
+            + set_date.seconds as u32;
+
+        Ok((date as u32, time as u32))
+    }
+
+    fn call_driver(&self, command: DateTimeCommand, r2: usize, r3: usize) -> CommandReturn {
+        match command {
+            DateTimeCommand::ReadDateTime => {
+                let date_result = self.date_time.get_date_time();
+                match date_result {
+                    Result::Ok(()) => {
+                        self.in_progress.set(true);
+                        CommandReturn::success()
+                    }
+                    Result::Err(e) => CommandReturn::failure(e),
```

Comment:
This potentially needs to generate an upcall. If a read was queued after a set, and when this read happens an error occurs, then the app calling get will never get a callback. But, if nothing was queued, then this can return an error directly in response to the command syscall.

Reply:
This is true of all command return errors in this function.